src/index.js

```diff
@@ -0,0 +1,864 @@
+            this.control.delete(peer)
+        }
+    }
+
+    /**
+     * When a peer has dialed into another peer, it sends its subscriptions to it.
+     *
+     * @param {PeerInfo} peerInfo
+     * @param {Connection} conn
+     * @param {Function} callback
+     *
+     * @returns {void}
+     *
+     */
+    _onDial (peerInfo, conn, callback) {
+        super._onDial(peerInfo, conn, (err) => {
+            if (err) return callback(err)
+            const idB58Str = peerInfo.id.toB58String()
+            const peer = this.peers.get(idB58Str)
+            if (peer && peer.isWritable) {
+                // Immediately send my own subscription to the newly established conn
+                peer.sendSubscriptions(this.subscriptions)
+            }
+            nextTick(() => callback())
+        })
+    }
+
+    /**
+     * Processes a peer's connection to another peer.
+     *
+     * @param {String} idB58Str
+     * @param {Connection} conn
+     * @param {Peer} peer
+     *
+     * @returns {void}
+     *
+     */
+    _processConnection (idB58Str, conn, peer) {
+        pull(
+            conn,
+            lp.decode(),
+            pull.map((data) => RPC.decode(data)),
+            pull.drain(
+                (rpc) => this._onRpc(idB58Str, rpc),
+                (err) => this._onConnectionEnd(idB58Str, peer, err)
+            )
+        )
+    }
+
+    /**
+     * Handles an rpc request from a peer
+     *
+     * @param {String} idB58Str
+     * @param {Object} rpc
+     * @returns {void}
+     */
+    _onRpc (idB58Str, rpc) {
+        if(!rpc){
+            return
+        }
+        //console.log(rpc)
+        this.log('rpc from', idB58Str)
+        const controlMsg = rpc.control
+        const subs = rpc.subscriptions
+        const msgs = rpc.msgs
+
+        if (subs && subs.length) {
+            const peer = this.peers.get(idB58Str)
+            if (peer) {
+                peer.updateSubscriptions(subs)
+                this.emit('meshsub:subscription-change', peer.info, peer.topics, subs)
+            }
+
+            subs.forEach((subOptMsg) => {
+                let t = subOptMsg.topicID
+                let topicSet = this.topics.get(t)
+                if (subOptMsg.subscribe) {
+                    if (!topicSet) {
+                        /**
+                         * @type Set<Peer>
+                         */
+                        topicSet = new Set()
+                        this.topics.set(t, topicSet.add(peer))
+                    }
+                } else {
+                    if (!topicSet) {
+                        return
+                    }
+                    this.topics.set(t, topicSet.delete(peer))
+                }
+            })
+        }
+
+        if (msgs && msgs.length) {
+            this._processRpcMessages(utils.normalizeInRpcMessage(msgs))
+        }
+
+        if (!controlMsg) {
+            return
+        }
+
+        let iWant = this._handleIHave(idB58Str, controlMsg)
```

Comment:
It might make sense to pass in the array of each control message type here.
eg: `this._handleIHave(peer, controlMsg.ihave)`